** PROJECT 1. DATA ANALYSIS PROJECT **

**Note:** *The purpose of this project is to analyze data from Google Play and Apple Store, and identify which application genres attract large number of users in each platform. By determining which genre attracts more users, we can recommend businesses to post their ads on those genre of applications to attract more viewers.*

*In this project, I will determine the most popular app genre by three criteria.*

*First, popular app genre is determined by the number of existing apps in each genre. This criterion works on both ios and apple dataset.*

*Second, popular app genre is determined by the number of reviews in each genre. Large number of reviews can imply there active users in these apps. Therefore, it is a good place to post the ads. I will use this criterion on Apple Store data.*

*Third, popular app genre is determined by the number installs in each genre. I will use this criterion on Google Play data.*

** 1. OPENING .CSV FILE**

**Note:** *Let's open the file using open(), reader() and list() functions*

In [1]:
from csv import reader
path = "AppleStore.csv"
opened_file = open(path)
read_file = reader(opened_file)
ios = list(read_file)
print(ios[:2])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']]


In [2]:
from csv import reader
path = "googleplaystore.csv"
opened_file = open(path)
read_file = reader(opened_file)
android = list(read_file)
print(android[:2])

[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']]


** 2. EXPLORING THE DATA**

**Note:** *Let's write a function that displays our data in a readable way and explore our two dataset*

In [3]:
def explore_data(dataset, start, end, rows_and_columns = False):
    """A function that displays dataset in an easily readable way.
    
    Args:
        dataset(list of lists): any dataset
        start(integer): row index to start with
        end(integer): row index to end with
        rows_and_columns(boolean): set as True if you want to see 
                                   the number of rows and columns 
                                   in your dataset 
    Returns: 
        list: lists by a line
        
    """
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [4]:
explore_data(ios, 1, 5, rows_and_columns = True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7198
Number of columns: 16


In [5]:
explore_data(android, 1, 5, rows_and_columns = True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10842
Number of columns: 13


In [6]:
print(ios[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [7]:
print(android[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


** 3. DATA CLEANING**

**3.1. Delete the row with wrong data entry:** *There is a wrong data entry in rows 10471 to 10473 in android data where Category column has number rather than words. Let's find that row and delete it from our dataset.* 

In [8]:
print(android[0])
print("\n")
print(android[10471])
print("\n")
print(android[10472])
print("\n")
print(android[10473])
print("\n")

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Jazz Wi-Fi', 'COMMUNICATION', '3.4', '49', '4.0M', '10,000+', 'Free', '0', 'Everyone', 'Communication', 'February 10, 2017', '0.1', '2.3 and up']


['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']




In [9]:
del android[10473]
print(android[10473])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


**3.2. Count duplicated data entries:** *Loop through app names to count how many apps are duplicated versus how many are unique.*

In [10]:
duplicate_apps = []
unique_apps =[]

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('Number of duplicated android apps:',len(duplicate_apps))
print('Number of unique android apps:', len(unique_apps))

Number of duplicated android apps: 1181
Number of unique android apps: 9660


In [11]:
duplicate_apps = []
unique_apps =[]

for app in ios:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('Number of duplicated ios apps:',len(duplicate_apps))
print('Number of unique ios apps:', len(unique_apps))

Number of duplicated ios apps: 0
Number of unique ios apps: 7198


**3.3. Choose deletion criterion for duplicated rows:** *Print out some duplicate apps to choose your criterion to delete some entries*

In [12]:
print(android[0])
for app in android:
    name = app[0]
    if name == 'Facebook':
        print(app)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']


**3.4. Delete duplicated rows based on the criterion:** *Within the duplicated entries we can use number of reviews as a criterion to delete duplicates. The higher the number of reviews, the more recent the entry is.*

**3.4.1.** *Let's write a dictionary for maximum number of reviews*

In [13]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = app[3]
    
    if name in reviews_max and float(reviews_max[name]) < float(n_reviews):
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

**3.4.2.** *Let's loop through each application and keep the entries with highest number of reviews for duplicated data. Finally, check whether we have 9660 rows including the header.*

In [14]:
android_unique = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = app[3]
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_unique.append(app)
        already_added.append(name)

In [15]:
print(len(android_unique))

9660


**3.5. Delete non-english apps:** *Each character we use in a string has a corresponding number associated with it. We can get the corresponding number of each character using the ord() built-in function to identify English words. English text characters range from 0 to 127.*

**3.5.1.** *Let's write a function that loops through each name and see if there is any character with corresponding number above 127.*

In [16]:
def detect_eng(word):
    """A function that detects strings written in English
    
    Args:
        word(string): strings you want to check whether it's 
                      written in English
    Returns:
        boolean: True if the word is in English and False if not
    """
    for i in word:
        if ord(i) > 127:
            return False
    return True
print(detect_eng('Instagram'))
print(detect_eng('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(detect_eng('Docs To Go™ Free Office Suite'))
print(detect_eng('Instachat 😜'))

True
False
False
False


**3.5.2.** *Above function cannot detect English names with emoji and symbols. Therefore, let's modify the function that it returns True if we have less than or equal to 3 characters outside the range of 127.*

In [17]:
def detect_eng(word):
    """A fuction that detects strings written in English but also allows
       less or equal to three non-english strings such as emoji
       
       Args:
           word(string): strings you want to check whether it's written 
                         in English
       Returns:
           boolean: True if the word is English and has less than 3 non-english
                    strings and False if not. 
    """
    non_ascii = 0
    
    for character in word:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

print(detect_eng('Instagram'))
print(detect_eng('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(detect_eng('Docs To Go™ Free Office Suite'))
print(detect_eng('Instachat 😜'))

True
False
True
True


**3.5.3.** *Using the function we wrote, let's extract English apps from each platform and store it as a list of lists.*

In [18]:
android_english = []
for app in android_unique:
    if detect_eng(app[0]) == True:
        android_english.append(app)
print('Number of english android apps:',len(android_english))

Number of english android apps: 9615


In [19]:
ios_english = []
for app in ios:
    if detect_eng(app[0]) == True:
        ios_english.append(app)
print('Number of english ios apps:', len(ios_english))

Number of english ios apps: 7198


**3.6. Isolate free apps:** *Some apps are free and some are not free let's isolate free apps because price can affect apps' popularity.*

In [20]:
android_free = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_free.append(app)
        
print('Number of free android apps:',len(android_free))

Number of free android apps: 8864


In [21]:
ios_free = []

for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_free.append(app)
        
print('Number of free ios apps:',len(ios_free))

Number of free ios apps: 4056


**4. FINDING THE MOST POPULAR APP GENRE/CATEGORY BY NUMBER OF EXISTING APPS**

**4.1. Identifying columns to compare:** *Let's see which column can be used to generate frequency tables to find out what are the most common genres in each market*

In [22]:
print(ios[0])
print(android[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


**Note:** *Following columns could be used from each application data to find out what are the most common genres in each market: prime_genre(ios column 11), Genres(android column 9), Category(android column 1])*

In [23]:
print(ios[:2])
print(android[:2])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']]
[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']]


**4.2. Creating frequency table:** *Let's create a table that shows the number of apps in each genre/category by percentage*

**4.2.1.** *Let's write a function that takes a dataset and column index as inputs and creates frequency table as a dictionary*

In [24]:
def freq_table(dataset,index):
    """A function that prepares frequency table for any column.
    
    Args:
        dataset(list of lists): cleaned ios or android application data
        index(integer): column index you want to see the frequency
    
    Returns:
        dictionary: keys are unique values and values are frequencies 
                    represented by percentage 
    """
    table = {}
    total = 0
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    table_percent = {}
    for key in table:
        percentage = (table[key]/total)*100
        table_percent[key] = percentage
    return table_percent

**4.2.2.** *Let's write a function that takes dictionary and displays it as a table (In order to display data in an ordered way, I need to transform the dictionary to tuple and use sorted() function. Otherwise, it is hard to order data in dictionary form)*

In [25]:
def display_table(dataset, index):
    """A function that displays ranked frequency table
    
    Args:
        dataset(list of lists): cleaned ios or android application dataset
        index(integer): column index you want to see the averages w.r.t 
                        each genre
                        
    Functions used inside:
        freq_table(dataset,index): turns data into frequency table in 
                                   dictionary type
        
    Returns:
        string: displays dictionary values ranked by the values
        
    """
    
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

**4.2.3.** *Let's generate frequency tables for the columns prime_genre, Genres, and Category*

*Prime genre in ios application data:*

In [26]:
print(display_table(ios_free,11))

Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032
None


**Explanation:** *English and free apps in ios app store is dominated by games (~56%) followed by entertainment (~8%) and least number of apps are medical apps (~0.2%).*

*Overall, entertainment type of apps (games, entertainment, photo, social media, etc.) while more practical apps (education, business, medical, etc.) are least common.*

*However, we need to see the number of users to identify which apps are more popular because above category percentage table cannot tell us which ones are commonly used in the market.*

*Genres in android application:*

In [27]:
print(display_table(android_free,9))

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

**Explanation:** *English and free apps in android app store is not as dominated as ios store by one genre. The most common apps on android apps market is tools (~8%) followed by entertainment (~6%) and least number of apps are adventure and education (-.01%). However, we also have to mention that genres in android app seems very repetitive compared to ios apps market. For instance, we can see genres like education, entertainment, fitness, music, and etc in many other different genres.*

*Category in android:*

In [28]:
print(display_table(android_free,1))

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

**Explanation:** *If we look at android apps by category, it tells a different story. English and free apps in android apps store is dominated by family (~19%) and games (~10%) apps.*

*It's hard to tell the difference between the genres and the categories in Google Play and as I mentioned before, Google Play genres are very repetitive. Therefore, we cannot conclude that Google Play has more balanced landscape of both practical and entertainment apps."*

**5. FINDING THE MOST POPULAR APPLE STORE APPS BY USER RATINGS**

**Note:** *In this section, I will rank app genres by number of reviews. Having many number of reviews shows how active the users are in those genres of apps. Therefore, it might be good place for companies to post their ads. I will apply this criterion on Apple Store apps.*

**5.1. Find which column has total number of reviews:** *Column 5 for Apple Store data contains the number of reviews.* 

In [29]:
print(ios[0])
print('\n')
print(ios_free[1])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


**5.2. Function for genre averages:** *Let's write a function that takes column_number as an input and finds that column averages with respect to each genre* 

In [30]:
def genre_averages(dataset, index, genre_index):
    """Calculates average values for each genre
    
    Args:
        dataset(list of lists): cleaned ios or android application dataset
        index(integer): column index to see averages w.r.t each genre 
        genre_index(integer): column index for genre 
        
    Returns:
        strings: groups the column by genre and prints string o f.....
        
    """
    genre_dict = freq_table(dataset, genre_index)
    table = {}

    for genre in genre_dict:
        genre_sum = 0 
        genre_len = 0
        for row in dataset:
            value = float(row[index])
            if row[genre_index] == genre:            
                row_value = float(row[index])
                genre_sum += row_value
                genre_len += 1
        if genre_len != 0 and genre_sum != 0:
            table[genre] = round(genre_sum/genre_len,2) 
    return table

**5.3. Function to display genre rating averages by ranking:** *Since it is hard to rank data written in dictionary form, I'll write another function to transform the data into tuple and prints data by rank. This function takes dataset, column index to calculate averages as an input and displays the data by rank. I'll use the genre_averages() function that I've just written above*

In [31]:
def display_table_genre_avg(dataset, index, genre_index):
    """Displays ranked averages w.r.t genres
    
    Args:
        dataset(list of lists): cleaned ios or android application dataset
        index(integer): column index you want to see the averages w.r.t each genre
    
    Results:
        NonType: prints out each list as NonType 
    """
    table = genre_averages(dataset, index, genre_index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [32]:
display_table_genre_avg(ios_free, 5, 11)

Reference : 67447.9
Music : 56482.03
Social Networking : 53078.2
Weather : 47220.94
Photo & Video : 27249.89
Navigation : 25972.05
Travel : 20216.02
Food & Drink : 20179.09
Sports : 20128.97
Health & Fitness : 19952.32
Productivity : 19053.89
Games : 18924.69
Shopping : 18746.68
News : 15892.72
Utilities : 14010.1
Finance : 13522.26
Entertainment : 10822.96
Lifestyle : 8978.31
Book : 8498.33
Business : 6367.8
Education : 6266.33
Catalogs : 1779.56
Medical : 459.75


**Explanation:** *We can see that for Apple Store apps 'Reference','Music', and 'Social Networking' are the most reviewed genres ; therefore, might be a good place to post some ads.

**6. FINDING THE MOST POPULAR GOOLGE PLAY APPS BY NUMBER OF INSTALLS**

**6.1. Frequency table for installs:** *As we can see installs column doesn't have a proper numbers. It has "+" and ",". Therefore, we need to eliminate these symbols first.*

In [33]:
display_table(android_free, 5) 

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


**6.2. Find the number of installs by genre:** *Let's loop over the unique genres of the Google Play data set and count the average number of installs.*

In [34]:
def genre_install_averages(dataset, install_index, genre_index):
    """Calculates average installs for each genre
    
    Args:
        dataset(list of lists): cleaned ios or android application dataset
        index(integer): column index for number of installs column 
        genre_index(integer): column index for genre 
        
    Returns:
        dictionary: keys are genres and values are average installs
        
    """
    genre_dict = freq_table(dataset, genre_index)
    table = {}
    
    for genre in genre_dict:
        total = 0
        len_genre = 0
        for row in dataset:
            app_genre = row[genre_index]
            if app_genre == genre:            
                n_installs = row[install_index]
                n_installs = n_installs.replace(',', '')
                n_installs = n_installs.replace('+', '')
                total += float(n_installs)
                len_genre += 1
        table[genre] = round(total/len_genre,2)
    return table

In [35]:
print(genre_install_averages(android_free, 5, 1))

{'VIDEO_PLAYERS': 24727872.45, 'GAME': 15588015.6, 'LIBRARIES_AND_DEMO': 638503.73, 'ENTERTAINMENT': 11640705.88, 'COMMUNICATION': 38456119.17, 'FOOD_AND_DRINK': 1924897.74, 'TRAVEL_AND_LOCAL': 13984077.71, 'SPORTS': 3638640.14, 'NEWS_AND_MAGAZINES': 9549178.47, 'PRODUCTIVITY': 16787331.34, 'HEALTH_AND_FITNESS': 4188821.99, 'PHOTOGRAPHY': 17840110.4, 'LIFESTYLE': 1437816.27, 'COMICS': 817657.27, 'FAMILY': 3695641.82, 'MAPS_AND_NAVIGATION': 4056941.77, 'BOOKS_AND_REFERENCE': 8767811.89, 'EDUCATION': 1833495.15, 'EVENTS': 253542.22, 'BUSINESS': 1712290.15, 'SOCIAL': 23253652.13, 'PARENTING': 542603.62, 'TOOLS': 10801391.3, 'FINANCE': 1387692.48, 'PERSONALIZATION': 5201482.61, 'MEDICAL': 120550.62, 'AUTO_AND_VEHICLES': 647317.82, 'BEAUTY': 513151.89, 'DATING': 854028.83, 'WEATHER': 5074486.2, 'ART_AND_DESIGN': 1986335.09, 'SHOPPING': 7036877.31, 'HOUSE_AND_HOME': 1331540.56}


print(android_free[:10

In [36]:
def display_genre_install_avg(dataset, install_index, genre_index):
    """Displays ranked installs averages w.r.t genres
    
    Args:
        dataset(list of lists): cleaned ios or android application dataset
        install_index(integer): column index for number of installs data
        genre_index(integer): column index for genre data
        
    Results:
        NonType: prints out each list as NonType 
    """
    table = genre_install_averages(dataset, install_index, genre_index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [37]:
print(display_genre_install_avg(android_free, 5, 1))

COMMUNICATION : 38456119.17
VIDEO_PLAYERS : 24727872.45
SOCIAL : 23253652.13
PHOTOGRAPHY : 17840110.4
PRODUCTIVITY : 16787331.34
GAME : 15588015.6
TRAVEL_AND_LOCAL : 13984077.71
ENTERTAINMENT : 11640705.88
TOOLS : 10801391.3
NEWS_AND_MAGAZINES : 9549178.47
BOOKS_AND_REFERENCE : 8767811.89
SHOPPING : 7036877.31
PERSONALIZATION : 5201482.61
WEATHER : 5074486.2
HEALTH_AND_FITNESS : 4188821.99
MAPS_AND_NAVIGATION : 4056941.77
FAMILY : 3695641.82
SPORTS : 3638640.14
ART_AND_DESIGN : 1986335.09
FOOD_AND_DRINK : 1924897.74
EDUCATION : 1833495.15
BUSINESS : 1712290.15
LIFESTYLE : 1437816.27
FINANCE : 1387692.48
HOUSE_AND_HOME : 1331540.56
DATING : 854028.83
COMICS : 817657.27
AUTO_AND_VEHICLES : 647317.82
LIBRARIES_AND_DEMO : 638503.73
PARENTING : 542603.62
BEAUTY : 513151.89
EVENTS : 253542.22
MEDICAL : 120550.62
None


**Explanation:** *We can see that for Google Play apps 'Communication','Video_players', and 'Social' are the most installed genres; therefore, it might be a good place to post some ads.* 

*I have to note here that we definitely need to analyze the data further to confidently say which application genre attracts more people; thus, a good place to post ads on them. This analysis is rather rudimentary and for further project development we need to see many things such as checking the outliers or cleaning Google Play data further to incorporate user reviews, etc.*